### PR TITLE
Add Soft Actor-Critic (SAC) 

### PR DESCRIPTION
**Pull request by brahimdriss**
## Description

This PR introduces SAC, following the original article, for continuous action spaces.
The current implementation was evaluated on gym Pendulum and Mujoco Hopper (v2), "solving" both environments.

## Todo
- \[x] Switch to gymnasium
- \[x] Add seeding
- \[x] Handle multiple fitting
- \[x] Add benchmarks (Mujoco and classic gym control)

**Original articles:**
[Soft Actor-Critic: Off-Policy Maximum Entropy Deep Reinforcement Learning with a Stochastic Actor](https://arxiv.org/abs/1801.01290)
[Composable Deep Reinforcement Learning for Robotic Manipulation](https://arxiv.org/abs/1803.06773)
[Soft Actor-Critic Algorithms and Applications](https://arxiv.org/abs/1812.05905)

**Reference resources:**
[haarnoja/sac](https://github.com/haarnoja/sac)
[haarnoja/softqlearning](https://github.com/haarnoja/softqlearning)
[openai/spinningup](https://github.com/openai/spinningup/tree/master/spinup/algos/tf1/sac)
[vwxyzjn/cleanrl](https://github.com/vwxyzjn/cleanrl)
[DLR-RM/stable-baselines3](https://github.com/DLR-RM/stable-baselines3/tree/master/stable_baselines3/sac)

## Checklist
- \[x] My code follows the style guideline
To check :
   black --check examples rlberry *py
   flake8 --select F401,F405,D410,D411,D412 --exclude=rlberry/check_packages.py --per-file-ignores="__init__.py:F401",
- \[x] I have commented my code, particularly in hard-to-understand areas,
- \[x] I have made corresponding changes to the documentation,
- \[x] I have added tests that prove my fix is effective or that my feature works,
- \[x] New and existing unit tests pass locally with my changes,
- \[x] If updated the changelog if necessary,
- \[x] I have set the label "ready for review" and the checks are all green.

### CODE REVIEW

**Comment by brahimdriss** [comment]
Pendulum results on 10 runs. Not sure about the style of the plot, but it still can be change later.

sac_pendulum

MuJoCo results are next.

**Comment by brahimdriss** [comment]
### **MuJoCo results:**

**Hopper-v2 :** 1M timesteps - 5 runs 
Training episodic return : **2247.09 ± 929.56**
sac_hopper_v2

File: rlberry/agents/torch/sac/sac.py [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+
+        # initialize episode, steps and time counters
+        self.total_episodes = 0
+        self.total_timesteps = 0
+        self.time = time.time()
+
+    def policy(self, state):
+        assert self.cont_policy is not None
+        state = np.array([state])
+        state = torch.FloatTensor(state).to(self.device)
```
Comment (riiswa) [inline rlberry/agents/torch/sac/sac.py:228]: Just a nit, but there is some built-in Pytorch method to convert a np array to a torch tensor (`torch.as_tensor`, `torch.from_numpy`). I think `torch.as_tensor` allow you to specify the device directly.

File: rlberry/agents/torch/sac/sac.py [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+        )
+
+        with torch.no_grad():
+            # Select action using the current policy
+            next_state_actions, next_state_log_pi = self._select_action(
+                next_state.detach().cpu().numpy()
+            )
+            # Compute the next state's Q-values
+            q1_next_target = self.q1_target(
+                torch.cat([next_state, next_state_actions], dim=1)
```
Comment (riiswa) [inline rlberry/agents/torch/sac/sac.py:402]: Is this compatible with multidimensional shapes observations like images ? If not I suggest to use dim=-1

Comment (brahimdriss) [reply rlberry/agents/torch/sac/sac.py:402]: I think you are right, I'll change it to `dim=-1`.

File: rlberry/agents/torch/sac/sac.py (hunk 1 above)
Comment (brahimdriss) [reply rlberry/agents/torch/sac/sac.py:228]: Thanks for the suggestions Waris,  I tried here to be consistent with the other implementations of Rlberry, `FloatTensor` was used in some of them, but so was `torch.from_numpy`. That's why I kept it. Is there any difference in terms of performance otherwise ?

**Comment by KohlerHECTOR** [comment]
It seems that something is broken with the doc in the sac branch. When trying to change the api.rst , it breaks readthedocs compilation. So I reverted. But otherwise, algorithmically speaking, SAC looks good to me.
